lib/templateFactory.js

```diff
@@ -80,33 +81,72 @@ class TemplateFactory extends BaseFactory {
-     * @param  {String}     config.name          The template name
-     * @param  {String}     config.version       Version of the template
-     * @param  {String}     [config.label]       Label of template
-     * @return {Promise}                         Resolves template model or undefined if not found
+     * @param  {String}     fullTemplateName    Name of the template and the version or tag (e.g. chef/publish@1.2.3)
+     * @return {Promise}                        Resolves template model or undefined if not found
      */
-    getTemplate(config) {
-        // get all templates with the same name
-        return super.list({ params: { name: config.name } })
-            .then((templates) => {
-                // get templates that have version prefix as config.version
-                const filtered = templates.filter(template =>
-                    template.version.startsWith(config.version));
-
-                // Sort templates by descending order
-                filtered.sort((a, b) => compareVersions(b.version, a.version));
-
-                // If no config.label, return latest one
-                if (filtered.length > 0 && !config.label) {
-                    return filtered[0];
-                }
+    getTemplate(fullTemplateName) {
+        const match = TEMPLATE_NAME_REGEX.exec(fullTemplateName);
+        const templateName = match[TEMPLATE_NAME_MATCH];
+        const versionOrTag = match[VERSION_OR_TAG_MATCH];
+        const isVersion = VERSION_REGEX.exec(versionOrTag);
+
+        return new Promise((resolve) => {
+            // If no version or tag provided, return version is null
+            if (!versionOrTag) {
+                return resolve(null);
+            }
+
+            // If tag is passed in, get the version from the tag
+            if (!isVersion) {
+                // Lazy load factory dependency to prevent circular dependency issues
+                // eslint-disable-next-line global-require
+                const TemplateTagFactory = require('./templateTagFactory');
+                const templateTagFactory = TemplateTagFactory.getInstance();
+
+                return templateTagFactory.get({
+                    name: templateName,
+                    tag: versionOrTag
+                })
+                .then((templateTag) => {
+                    // Return -1 if no template tag exists
+                    if (!templateTag) {
+                        return resolve(-1);
+                    }
 
-                // Loop through filtered to find the first match for config.label
-                return filtered.find(template => template.labels.includes(config.label));
-            });
+                    return resolve(templateTag.version);
+                });
+            }
+
+            return resolve(versionOrTag);
+        })
+        .then((version) => {
+            // If no template tag exists, return undefined
+            if (version === -1) {
+                return undefined;
```

Comment:
I was looking at the original README and it said `undefined`, so I was trying to be consistent. I can change it to `null` if that's better.